File: code/client-side-impl/src/application/fileIO/GameLibraryIO.java
package application.fileIO;

import application.model.local_impl.game.Game;
import application.model.local_impl.game.GameLibrary;
import application.model.local_impl.game.Genre;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Utility class for parsing CSV data into a GameLibrary object. This class
 * provides functionality to read game information from a CSV-formatted string,
 * convert each line of the CSV into a Game object, and then collect these games
 * into a GameLibrary.
 * 
 * @author Thomas Lamont
 * @version Sprint 1
 */
public class GameLibraryIO {

	private static final String DATABASE_FILENAME = "merged_steam_game_database.csv";
	// private static final String DATABASE_FILENAME = "testData.csv";

	/**
	 * Parses a CSV string containing game data and returns a populated GameLibrary
	 * instance.
	 * 
	 * @return A GameLibrary containing all the games parsed from the CSV data.
	 */
	public static GameLibrary parseGamesFromFile() {
		GameLibrary library = new GameLibrary();
		Path path = Paths.get(DATABASE_FILENAME);

		try {
			List<String> lines = Files.readAllLines(path);

			for (int i = 1; i < lines.size(); i++) {
				String line = lines.get(i);
				List<String> fields = parseCsvLine(line);
				Game game = parseGame(fields);
				if (game != null) {
					library.addGame(game);
				}
			}
		} catch (IOException e) {
			System.err.println("Error reading CSV file: " + e.getMessage());
		}
		System.out.println(library.size());
		return library;
	}

	/**
	 * Parses a CSV string containing game data and returns a populated GameLibrary
	 * instance.
	 * 
	 * @param csvText The CSV data as a string, where each line represents a game.
	 * @return A GameLibrary containing all the games parsed from the CSV data.
	 */
	public static GameLibrary parseGamesFromText(String csvText) {
		GameLibrary library = new GameLibrary();
		String[] lines = csvText.split("\n");

		for (int i = 1; i < lines.length; i++) {
			String line = lines[i];
			List<String> fields = parseCsvLine(line);
			Game game = parseGame(fields);
			if (game != null) {
				library.addGame(game);
			}
		}

		return library;
	}

	private static List<String> parseCsvLine(String csvLine) {
		List<String> fields = new ArrayList<>();
		boolean inQuotes = false;
		StringBuilder buffer = new StringBuilder();

		for (char currChar : csvLine.toCharArray()) {
			if (currChar == '"') {
				inQuotes = !inQuotes;
			} else if (currChar == ',' && !inQuotes) {
				fields.add(buffer.toString());
				buffer = new StringBuilder();
			} else {
				buffer.append(currChar);
			}
		}
		fields.add(buffer.toString());

		return fields;
	}

	private static Game parseGame(List<String> fields) {
		try {
			int appId = Integer.parseInt(fields.get(0));
			String name = fields.get(1).replaceAll("^\"|\"$", "");
			String[] releaseDateParts = fields.get(2).split("/");
			int releaseDateMonth = Integer.parseInt(releaseDateParts[0]);
			int releaseDateYear = Integer.parseInt(releaseDateParts[2]);
			String developer = fields.get(3);
			List<Genre> genres = Arrays.stream(fields.get(7).split(";")).map(GameLibraryIO::toGenre)
					.filter(Optional::isPresent).map(Optional::get).collect(Collectors.toList());
			int positiveRatings = Integer.parseInt(fields.get(9));
			int negativeRatings = Integer.parseInt(fields.get(10));
			int averagePlaytime = Integer.parseInt(fields.get(11));
			String gamePhotoLink = fields.get(13);
			String description = fields.get(14);
			Game game = new Game(name, genres, appId, developer, releaseDateYear, releaseDateMonth, positiveRatings,
					negativeRatings, averagePlaytime, gamePhotoLink, description);
			// System.out.println("added" + game);
			return game;
		} catch (Exception e) {
			System.err.println("Error parsing game from CSV fields: " + fields);
			return null;
		}
	}

	private static Optional<Genre> toGenre(String genreStr) {
		String normalized = genreStr.toUpperCase().replace(" ", "_").replace("-", "_");
		try {
			return Optional.of(Genre.valueOf(normalized));
		} catch (IllegalArgumentException e) {
			return Optional.of(Genre.MISSING_GENRE);
		}
	}
}

File: code/server-side-impl/src/src/model/games/gameLibraryIO.py
'''
Created on Mar 2, 2024

@author: Thomas
'''
import csv
from src.model.games.game import Game, Genre
from src.model.games.gamelibrary import GameLibrary


class GameLibraryIO:
    """
    A class to handle Input/Output operations for a game library, specifically
    reading game details from a CSV file and parsing them into Game objects.
    
    Methods:
        parse_csv_line(csv_line): Parses a single line of CSV, considering the possibility of fields enclosed in quotes.
        to_genre(genre_str): Converts a genre string to a Genre enum, handling replacements for spaces and hyphens.
        parse_game(fields): Parses an array of fields into a Game object.
        parse_games_from_file(filename): Reads a CSV file and returns a GameLibrary object populated with games.
    """
    @staticmethod
    def parse_csv_line(csv_line):
        """
        Parses a single line from a CSV file, handling fields that may be enclosed in quotes and contain commas.
        
        Args:
            csv_line (str): A line from a CSV file.
            
        Returns:
            list[str]: A list of fields extracted from the csv_line.
        """
        fields = []
        in_quotes = False
        buffer = ''
        for curr_char in csv_line:
            if curr_char == '"':
                in_quotes = not in_quotes
            elif curr_char == ',' and not in_quotes:
                fields.append(buffer)
                buffer = ''
            else:
                buffer += curr_char
        fields.append(buffer) 
        return fields

    @staticmethod
    def to_genre(genre_str):
        """
        Converts a genre string from the CSV file to a Genre enum value.
        
        Args:
            genre_str (str): The genre string from the CSV file.
            
        Returns:
            Genre: The corresponding Genre enum value, or MISSING_GENRE if the string does not match any known Genre.
        """
        try:
            return Genre[genre_str.upper().replace(" ", "_").replace("-", "_")]
        except KeyError:
            return Genre.MISSING_GENRE

    @staticmethod
    def parse_game(fields):
        """
        Parses an array of strings into a Game object. Expects fields to be in a specific order as per the CSV format.
        
        Args:
            fields (list[str]): An array of fields representing game attributes.
            
        Returns:
            Game: A Game object populated with the provided attributes, or None if parsing fails.
            
        Raises:
            Exception: If there is an error in parsing the fields into a Game object.
        """
        try:
            game_id = int(fields[0])
            name = fields[1].strip('"') 
            release_date_parts = fields[2].split("/")
            release_date_month = int(release_date_parts[0])
            release_date_year = int(release_date_parts[2])
            developers = fields[3]
            genres = [GameLibraryIO.to_genre(genre_str) for genre_str in fields[7].split(";")]
            positive_ratings = int(fields[9])
            negative_ratings = int(fields[10])
            average_playtime = int(fields[11])
            game_photo_link = fields[13]
            description = fields[14]
            return Game(name, genres, game_id, developers, release_date_year, release_date_month,
                        positive_ratings, negative_ratings, average_playtime, game_photo_link, description)
        except Exception as e:
            print(f"Error parsing game from CSV fields: {e}")
            return None

    @staticmethod
    def parse_games_from_file(filename):
        """
        Reads games from a specified CSV file and populates a GameLibrary object with Game instances.
        
        Args:
            filename (str): The path to the CSV file containing game details.
            
        Returns:
            GameLibrary: A GameLibrary object populated with games parsed from the CSV file.
        """
        library = GameLibrary()
        with open(filename, mode='r', encoding='utf-8') as csv_file:
            csv_reader = csv.reader(csv_file)
            next(csv_reader) 
            for row in csv_reader:
                game = GameLibraryIO.parse_game(row)
                if game:
                    library.add_game(game)
        return library